src/thermal.cpp

```diff
@@ -0,0 +1,289 @@
+
+  // Determine actual temperatures to read -- start by checking whether a
+  // temperature range was given, in which case all temperatures in the range
+  // are loaded irrespective of what temperatures actually appear in the model
+  std::vector<int> temps_to_read;
+  if (settings::temperature_range[1] > 0.0) {
+    for (const auto& T : temps_available) {
+      if (settings::temperature_range[0] <= T &&
+          T <= settings::temperature_range[1]) {
+        temps_to_read.push_back(std::round(T));
+      }
+    }
+  }
+
+  switch (settings::temperature_method) {
+  case TEMPERATURE_NEAREST:
+    // Determine actual temperatures to read
+    for (const auto& T : temperature) {
+
+      auto i_closest = xt::argmin(xt::abs(temps_available - T))[0];
+      auto temp_actual = temps_available[i_closest];
+      if (std::abs(temp_actual - T) < settings::temperature_tolerance) {
+        if (std::find(temps_to_read.begin(), temps_to_read.end(), std::round(temp_actual))
+            == temps_to_read.end()) {
+          temps_to_read.push_back(std::round(temp_actual));
+        }
+      } else {
+        std::stringstream msg;
+        msg << "Nuclear data library does not contain cross sections for "
+          << name_ << " at or near " << std::round(T) << " K.";
+        fatal_error(msg);
+      }
+    }
+    break;
+
+  case TEMPERATURE_INTERPOLATION:
+    // If temperature interpolation or multipole is selected, get a list of
+    // bounding temperatures for each actual temperature present in the model
+    for (const auto& T : temperature) {
+      bool found = false;
+      for (int j = 0; j < temps_available.size() - 1; ++j) {
+        if (temps_available[j] <= T && T < temps_available[j + 1]) {
+          int T_j = std::round(temps_available[j]);
+          int T_j1 = std::round(temps_available[j + 1]);
+          if (std::find(temps_to_read.begin(), temps_to_read.end(), T_j) == temps_to_read.end()) {
+            temps_to_read.push_back(T_j);
+          }
+          if (std::find(temps_to_read.begin(), temps_to_read.end(), T_j1) == temps_to_read.end()) {
+            temps_to_read.push_back(T_j1);
+          }
+          found = true;
+        }
+      }
+      if (!found) {
+        std::stringstream msg;
+        msg << "Nuclear data library does not contain cross sections for "
+          << name_ << " at temperatures that bound " << std::round(T) << " K.";
+        fatal_error(msg);
+      }
+    }
+  }
+
+  // Sort temperatures to read
+  std::sort(temps_to_read.begin(), temps_to_read.end());
+
+  auto n_temperature = temps_to_read.size();
+  kTs_.reserve(n_temperature);
+  data_.reserve(n_temperature);
+
+  for (auto T : temps_to_read) {
+    // Get temperature as a string
+    std::string temp_str = std::to_string(T) + "K";
+
+    // Read exact temperature value
+    double kT;
+    read_dataset(kT_group, temp_str.data(), kT);
+    kTs_.push_back(kT);
+
+    // Open group for temperature i
```

Comment:
Potential typo: May have meant "temperature T" to indicate the iterator variable?  `i` does not appear to be defined in this scope.